src/newton.jl

```diff
@@ -1,52 +1,72 @@
+##################################################
 ## Classical derivative-based, iterative, root-finding algorithms: Newton and Halley
-
-## Newton
-function newton_itr(f, fp, x0; 
-                    xtol=4*eps(), xtolrel=4*eps(), ftol=4*eps(),
-                    maxsteps=100, maxfnevals=100)
-    update =  (o) -> begin
-        xn = o.xn[end]
-        fxn = o.fxn[end]
-        fpxn = o.fp(xn)
-        incfn(o)
-
-        xn1 = xn - fxn / fpxn
-        fxn1 = o.f(xn1)
-        incfn(o)
-
-        push!(o.xn, xn1)
-        push!(o.fxn, fxn1)
-
-    end
-
-    x, fx = promote(float(x0), f(float(x0)))
-    out = ZeroType(f, fp, nothing, update, [x], [fx],
-                   xtol, xtolrel, ftol,:not_converged, 
-                   0, maxsteps, 1, maxfnevals)
+## Historic, we have derivative free versions of similar order
 
 
-    out
+## Newton
+type Newton <: UnivariateZeroMethod
 end
 
+callable_function(method::Newton, f) = FirstDerivative(f, D(f))
+
+function init_state{T}(method::Newton, fs, x0::T)
+    state = UnivariateZeroState(x0,
+                                x0 + typemax(Int),
+                                fs.f(x0),
+                                fs.f(x0),
+                                Nullable{Vector{T}}(),
+                                0,
+                                1,
+                                false,
+                                false,
+                                false,
+                                false,
+                                "")
+    state
+end
 
+function update_state{T}(method::Newton, fs, o::UnivariateZeroState{T}, options::UnivariateZeroOptions)
+    xn = o.xn1
+    fxn = o.fxn1
+    fpxn = fs.fp(xn)
 
-function newton_method{T}(f, fp, x0::T;
-                       xtol=4*eps(), xtolrel=4*eps(), ftol=4eps(),
-                       maxsteps::Int=100, maxfnevals=100,
-                       verbose::Bool=false)
+    if isissue(fpxn)
+        o.stopped=true
+        return
+    end
+    
+    xn1 = xn - fxn / fpxn
+    fxn1 = fs.f(xn1)
+    incfn(o)
+    
+    o.xn0, o.xn1 = xn, xn1
+    o.fxn0, o.fxn1 = fxn, fxn1
 
-    o = newton_itr(f, fp, x0;
-                  xtol=xtol, xtolrel=xtolrel, ftol=ftol,
-                  maxsteps=maxsteps, maxfnevals=maxfnevals)
 
+    incsteps(o)
+    
+end
 
-    for x in o
-        nothing
+## extra work to allow for complex values
+function derivative_free_setup{T<:AbstractFloat}(method::Newton, fs::CallableFunction, x0::Union{T, Complex{T}};
+                                  bracket=Nullable{Vector{T}}(),
+                                  xabstol=zero(T), xreltol=zero(T),
+                                  abstol=4*eps(T), reltol=4*eps(T),
+                                  maxevals=40, maxfnevals=typemax(Int),
+                                  verbose::Bool=false)
+    x = map(float, x0)
+
+    if isa(x, Complex)
+        bracket = Nullable{Vector{Complex{T}}}()     # bracket makes no sense for complex input, but one is expected
```

Comment:
This looks like a type instability. `bracket` will from the arguments be a `Nullable{Vector{T}}` but will here be a `Nullable{Vector{Complex{T}}}`. However, very little work is done with it here until the next function barrier so it is probably fine.